### PR TITLE
Add Q-Q plot to the ResidualsPlot (#853)

### LINKED ISSUE #853: Add Q-Q plot to the yellowbrick.regressor.residuals class
You already have a histogram feature for the `ResidulaPlot` method.
It will be immensely helpful to add a standard normality check method to the parent `yellowbrick.regressor.residuals` class like Q-Q plot.

Your functional interface is gearing up to be similar to statistical languages like **R** where you can throw the fitted model inside a function to generate more insight - mostly a visualization or a statistical score.

[Q-Q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot) ability will be one of such basic statistical insights that can add value.

### PR DESCRIPTION
**Pull request by VladSkripniuk**
This PR fixes #853 which requested a feature to allow the user to create Q-Q plot beside the residual plot as a mean of normality checking of the residuals.

I have made the following changes:

1. Plot Q-Q plot using ``scipy.stats.probplot``
2. Add tests.

### Sample Code and Plot

```
from yellowbrick.regressor.residuals import ResidualsPlot
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from yellowbrick.datasets import load_concrete

X, y = load_concrete()

X_train, X_test, y_train, y_test = train_test_split(X, y)
viz = ResidualsPlot(LinearRegression(), hist=False, qqplot=True)

viz.fit(X_train, y_train)
viz.score(X_test, y_test)
viz.show()

```
image

### CHECKLIST

- [x] _Is the commit message formatted correctly?_
- [ ] _Have you noted the new functionality/bugfix in the release notes of the next release?_

- [x] _Included a sample plot to visually illustrate your changes?_
- [x] _Do all of your functions and methods have docstrings?_
- [x] _Have you added/updated unit tests where appropriate?_
- [x] _Have you updated the baseline images if necessary?_
- [x] _Have you run the unit tests using `pytest`?_
- [x] _Is your code style correct (are you using PEP8, pyflakes)?_
- [ ] _Have you documented your new feature/functionality in the docs?_

- [ ] _Have you built the docs using `make html`?_

### CODE REVIEW

File: yellowbrick/regressor/residuals.py [hunk 1]
```diff
@@ -502,6 +504,7 @@ def __init__(
         model,
         ax=None,
         hist=True,
+        qqplot=False,
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: 1. qqplot parameter should be added to the docstring.
2. The docstring should contain a description of what qqplot 
3. The docstring should include what are the available options for qqplot (ie. True, False)

File: yellowbrick/regressor/residuals.py [hunk 2]
```diff
@@ -560,6 +579,26 @@ def hax(self):
 
         return hax
 
+    @memoized
+    def qqax(self):
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Great job with adding this here!

File: yellowbrick/regressor/residuals.py [hunk 3]
```diff
@@ -670,6 +709,12 @@ def draw(self, y_pred, residuals, train=False, **kwargs):
                 residuals, bins=50, orientation="horizontal", density=True, color=color
             )
 
+        # Add residuals histogram
+        if self.qqplot in {True}:
+            osm, osr = probplot(residuals, dist='norm', fit=False)
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: From my understanding probplot generates a probability plot, which should not be confused with a Q-Q or a P-P plot - are(osm,osr) the appropriate values to use for a Q-Q plot?????

File: yellowbrick/regressor/residuals.py [hunk 4]
```diff
@@ -670,6 +709,12 @@ def draw(self, y_pred, residuals, train=False, **kwargs):
...
+
+            self.qqax.scatter(osm, osr, c=color, alpha=alpha, label=label)
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Refer to comment above.

File: yellowbrick/regressor/residuals.py [hunk 5]
```diff
@@ -705,6 +750,12 @@ def finalize(self, **kwargs):
             self.hax.axhline(y=0, c=self.colors["line"])
             self.hax.set_xlabel("Distribution")
 
+        # Finalize the histogram axes
+        if self.qqplot:
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Refer to comment above

File: yellowbrick/regressor/residuals.py [hunk 6]
```diff
@@ -531,9 +534,25 @@ def __init__(
                 "False, 'density', or 'frequency'".format(hist)
             )
 
+        self.qqplot = qqplot
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Great job on handling qqplot parameter input.

File: tests/test_regressor/test_residuals.py [hunk 7]
```diff
@@ -280,6 +280,33 @@ def test_residuals_plot(self):
...
+        visualizer = ResidualsPlot(LinearRegression(), hist=False,
+                                                        qqplot=True, ax=ax)
+
+        visualizer.fit(self.data.X.train, self.data.y.train)
+        visualizer.score(self.data.X.test, self.data.y.test)
+        visualizer.finalize()
+
+        self.assert_images_similar(visualizer)
+
+    def test_either_hist_or_QQ_plot(self):
```
Comment (lwgray) [inline tests/test_regressor/test_residuals.py]: Great test to catch this error.

File: tests/test_regressor/test_residuals.py [hunk 8]
```diff
@@ -280,6 +280,33 @@ def test_residuals_plot(self):
 
         self.assert_images_similar(visualizer)
 
+    @pytest.mark.xfail(
+        IS_WINDOWS_OR_CONDA,
+        reason="font rendering different in OS and/or Python; see #892",
+    )
+    def test_residuals_plot_QQ_plot(self):
```
Comment (lwgray) [inline tests/test_regressor/test_residuals.py]: Once we get the `probplot` figured out, I am sure this test will run ok

File: yellowbrick/regressor/residuals.py (hunk 3 above)
Comment (VladSkripniuk) [reply yellowbrick/regressor/residuals.py]: `scipy.stats.probplot` computes quantiles of sample data and quantiles of specified theoretical distribution, which is exactly what is needed for Q-Q plot. So, the quote from the scipy documentation
> probplot generates a probability plot, which should not be confused with a Q-Q or a P-P plot.

is a bit misleading, because `scipy.stats.probplot` plots exactly Q-Q plot. I'm pretty sure, that I plotted Q-Q plot using `scipy.stats.probplot` correctly.
Here are some links to support my argument: [definition of probability plot](https://en.wikipedia.org/wiki/Probability_plot), [definition of Q-Q plot](https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot), [answer on stackoverflow](https://stackoverflow.com/a/13865874).

File: yellowbrick/regressor/residuals.py (hunk 1 above)
Comment (VladSkripniuk) [reply yellowbrick/regressor/residuals.py]: Thank you for pointing this out, I added qqplot to the docstring of ResidualsPlot.

**Comment by lwgray** [comment]
@VladSkripniuk For each visualizer there is a "Quick Method".  It allows for the Visualizer to be called on a single line.  Since you modified the ResidualsPlot Class you will need to alter the quick method also.  The quickmethod begins as [Line 766](https://github.com/DistrictDataLabs/yellowbrick/blob/71c57d136b4f24f77b375bf0f3797e9c7c1d3deb/yellowbrick/regressor/residuals.py#L766)

Tasks
1. Add the qqplot parameter [Lines 771 to 786](https://github.com/DistrictDataLabs/yellowbrick/blob/71c57d136b4f24f77b375bf0f3797e9c7c1d3deb/yellowbrick/regressor/residuals.py#L771-L786)
2. Add qqplot to the instantiation of ResidualsPlot [Lines 878 to 889](https://github.com/DistrictDataLabs/yellowbrick/blob/71c57d136b4f24f77b375bf0f3797e9c7c1d3deb/yellowbrick/regressor/residuals.py#L878-L889)
3. Add qqplot to the docstring

**Comment by lwgray** [comment]
Another, note... Your branch is out-of-date with the base branch... Please Update this branch before pushing any changes

File: yellowbrick/regressor/residuals.py (hunk 3 above)
Comment (lwgray) [reply yellowbrick/regressor/residuals.py]: There is a difference between probplot and Q-Q plot however, scipy's `probplot` actually provides the sample quantiles on the y-axis and Theorectical quantile on the x-axis which is exactly the definition of a Q-Q plot.  Just for our education, Normally, ProbaPlot is the same as a Q-Q plot, however probabilities are shown in the scale of the theoretical distribution (x-axis) and the y-axis contains unscaled quantiles of the sample data.

**Comment by VladSkripniuk** [comment]
@lwgray thank you for your help with this PR!

File: yellowbrick/regressor/residuals.py [hunk 9]
```diff
@@ -719,6 +776,7 @@ def residuals_plot(
     y_test=None,
     ax=None,
     hist=True,
+    qqplot=False,
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Thanks for this addition!

File: yellowbrick/regressor/residuals.py [hunk 10]
```diff
@@ -772,6 +830,12 @@ def residuals_plot(
         If set to 'density', the probability density function will be plotted.
         If set to True or 'frequency' then the frequency will be plotted.
 
+    qqplot : {True, False}, default: False
```
Comment (lwgray) [inline yellowbrick/regressor/residuals.py]: Thanks for adding this here.  We do our best to keep our docstring up-to-date because this is how our docs are generated.

File: docs/api/regressor/residuals.rst [hunk 11]
```diff
@@ -52,6 +52,19 @@ Note that if the histogram is not desired, it can be turned off with the ``hist=
 
 .. warning:: The histogram on the residuals plot requires matplotlib 2.0.2 or greater. If you are using an earlier version of matplotlib, simply set the ``hist=False`` flag so that the histogram is not drawn.
 
+Histogram can be replaced with a Q-Q plot, which is a common way to check that residuals are normally distributed. If the residuals are normally distributed, then their quantiles when plotted against quantiles of normal distribution should form a straight line. The example below shows, how Q-Q plot can be drawn with a ``qqplot=True`` flag. Notice that ``hist`` has to be set to ``False`` in this case.
```
Comment (lwgray) [inline docs/api/regressor/residuals.rst]: This is great explanation.  Awesome on notifying the user that hist has to be set to False

**Comment by lwgray** [comment]
Your PR is currently blocked because of PR #1050 which resolves some bugs in the test but yours will be merged once #1050 is merged

**Comment by lwgray** [comment]
We fixed the test failures.... i will do a final review and if test past then I will merge this PR.  I have updated your branch so do a pull before you make anymore commits

**Comment by lwgray** [comment]
@bbengfort or @rebeccabilbro Travis has  Image Comparison failures for the Q-Q plot related tests although they are passing locally.  I had a couple of questions:

1. The local test for Q-Q plot xfail.  Is it necessary to have the pytest xfail decorator with the Q-Qplot [test](https://github.com/DistrictDataLabs/yellowbrick/pull/1042/files#diff-4c54b87374c2d95f14a964bc29763bafR283-R286) ?

2. Would an appropriate fix be to clear the actual & base images, run residual plots tests, then move the actual images to the baseline images (ie)

python -m  tests.images -C tests/test_regressor/test_residuals.py
pytest -vs tests/test_regressor/test_residuals.py
python -m tests.images tests/test_regressor/test_residuals.py

File: tests/test_regressor/test_residuals.py [hunk 12]
```diff
@@ -280,6 +280,33 @@ def test_residuals_plot(self):
...
+        Image similarity of residuals and Q-Q plot on random data with OLS
+        """
+        _, ax = plt.subplots()
+
+        visualizer = ResidualsPlot(LinearRegression(), hist=False,
+                                                        qqplot=True, ax=ax)
+
+        visualizer.fit(self.data.X.train, self.data.y.train)
+        visualizer.score(self.data.X.test, self.data.y.test)
+        visualizer.finalize()
```
Comment (bbengfort) [inline tests/test_regressor/test_residuals.py:298]: @lwgray I recommend removing this call to finalize and seeing if that helps fix the tests on Travis - there is a known problem that the image similarity tests aren't removing text from the alternate axes (as you can see in the baseline image) - and text is one of the primary reasons we have a divergence between AppVeyor and Travis. 

Were the baseline images created on a Windows machine or Mac/Linux? That might also contribute to this; we have an xfail for windows or conda, but if the image were generated on Windows than it would fail anyway on Travis.

Comment (lwgray) [reply tests/test_regressor/test_residuals.py:298]: @VladSkripniuk can you remove the call to finalize as described in the above comment from @bbengfort ?
I updated your branch so make sure you pull before you commit and new changes.
Also, did you create the baseline images on Windows or Mac?

Comment (VladSkripniuk) [reply tests/test_regressor/test_residuals.py]: @lwgray Sure, I just removed the call to finalize. I created `tests/baseline_images/test_regressor/test_residuals/test_residuals_plot_QQ_plot.png` in commit 71f83bbf1ae314a5573e154aa7d92a24169816a3 using Ubuntu 18.04.

Comment (lwgray) [reply tests/test_regressor/test_residuals.py]: Hi @VladSkripniuk your baseline image is out-of-date.  We have since merged fixes issues we were having with matplotlib,  so you will need to recreate your baseline images.  You can perform the following commands to reset your baseline images.

python -m tests.images -C tests/test_regressor/test_residuals.py
pytest -vs tests/test_regressor/test_residuals.py
python -m tests.images tests/test_regressor/test_residuals.py

Comment (VladSkripniuk) [reply tests/test_regressor/test_residuals.py]: @lwgray I executed the commands you mentioned:
```
python -m tests.images -C tests/test_regressor/test_residuals.py
pytest -vs tests/test_regressor/test_residuals.py
python -m tests.images tests/test_regressor/test_residuals.py
```
 and ran `python -m pytest tests/test_regressor/test_residuals.py` and tests passed successfully on my machine. I upload the updated version of `tests/baseline_images/test_regressor/test_residuals/test_residuals_plot_QQ_plot.png`

Comment (lwgray) [reply tests/test_regressor/test_residuals.py]: @VladSkripniuk great call on upgrading mpl to 3.2.1 - I forgot to tell you to do that.

**Comment by lwgray** [comment]
@VladSkripniuk Thank you for your patience and your contribution.  I will start working on your other PRs.
